add a new command `opm clean`

pronan opened the pull request:
to delete folder and .gz file created by opm build or opm upload

agentzh commented:
@pronan You patch looks wrong to me. I don't think your code even runs:

```console
$ perl -c bin/opm
Global symbol "$root_dir" requires explicit package name (did you forget to declare "my $root_dir"?) at bin/opm line 2203.
Global symbol "$root_dir" requires explicit package name (did you forget to declare "my $root_dir"?) at bin/opm line 2204.
```

pronan commented:
@agentzh Sorry, chun ge, I fixed it. This time this commit  is fully tested and passed. `lua-resty-query` is just uploaded and cleaned via `sudo opm build && sudo opm upload && sudo opm dist-clean`

bungle commented:
It only cleans the latest. Should it clean the previously created versions as well?

pronan commented:
@bungle do you mean change `my $root_dir = "$dist_name-$version";` to `my $root_dir = "$dist_name-*";` so the final string will be like `rm -rf './lua-resty-datetime-*'` and`rm -rf './lua-resty-datetime-*.tar.gz'`?

bungle commented:
@pronan, I mean that maybe that should be considered as well. Not sure if that should be a default or should we introduce a command line argument.

```
-- clean latest or all?
opm dist-clean

-- clean current
opm dist-clean current      --or
opm dist-clean --current  -- or
opm dist-clean -c


-- clean all
opm dist-clean all       --or
opm dist-clean --all    -or
opm dist-clean -a
```

Or something?

agentzh commented:
@pronan Sorry, I changed my mind a bit. It seems like `opm clean dist` is better. So that we can also have `opm clean cache` in the future ;) What do you think?

I agree that `opm clean dist` should also clean up tarballs and directories belong to the older versions. But I do think @bungle's `current` argument or `--current` option are overkill.

I think it's good enough to just cleanup whatever looks like our leftover. Basically we can iterate through the current working directory, checking each file or directory, looking for

1. a directory whose name matches the pattern `/^\Q$name\E-[.\w]*\d[.\w]*$/` and also containing a file `dist.ini` in itself; and
2. a file whose name matches the pattern `/^\Q$name\E-[.\w]*\d[.\w]*\.tar\.gz$/`.

To iterate through the cwd, we can just use the `opendir`, `readdir`, and `closedir` sequence. We do have such examples in the existing `opm` code base :)

What's your opinion?

pronan commented:
@agentzh hello, chun ge. I agree with you.PR is tested and passed with your regex pattern.

agentzh commented on bin/opm:
```diff
@@ -2195,6 +2198,37 @@ sub rebase_path ($$$) {
+        opendir(my $dh, './') || die "Can't opendir";
+        my @dots = readdir($dh);
+        foreach my $x (@dots) {
+            if(-d $x){
```
Style: need a space after `if` and also before `{`. Please keep the coding style consistent with the rest of the code base. Thanks!

pronan replied:
done

agentzh commented on bin/opm:
```diff
@@ -2195,6 +2198,37 @@ sub rebase_path ($$$) {
+                    shell "rm -rf $x";
+                    say 'delete:',$x;
+                }
+            }else{
```
Style: need a space before and after `else`.

pronan replied:
done

agentzh commented on bin/opm:
```diff
@@ -2195,6 +2198,37 @@ sub rebase_path ($$$) {
+            if(-d $x){
+                if ($x =~ /^\Q$dist_name\E-[.\w]*\d[.\w]*$/){
+                    shell "rm -rf $x";
+                    say 'delete:',$x;
```
Better avoid using `say` to be consistent with the rest of the code base. Also need a space after the comma (`,`).

pronan replied:
done

agentzh commented on bin/opm:
```diff
@@ -2195,6 +2198,37 @@ sub rebase_path ($$$) {
+        my $data = read_ini($dist_file);
+        my $default_sec = $data->{default};
+        my $dist_name = $default_sec->{name};
+        opendir(my $dh, './') || die "Can't opendir";
```
Error message in case of `opendir` failures is not descriptive enough. Please see similar places in this file for examples.

pronan replied:
done

agentzh commented on bin/opm:
```diff
@@ -2195,6 +2198,37 @@ sub rebase_path ($$$) {
+    if (@_ == 0) {
+        err "no clean option specified.\n";
+    }
+    if ($_[0]  eq 'dist') {
```
Style: 2 spaces before `eq` while only one is needed.

agentzh replied:
Also, need a blank line before this `if` statement.

pronan replied:
done.

agentzh commented on bin/opm:
```diff
@@ -2195,6 +2198,37 @@ sub rebase_path ($$$) {
+        my $dist_file = "dist.ini";
+        my $data = read_ini($dist_file);
+        my $default_sec = $data->{default};
+        my $dist_name = $default_sec->{name};
```
Need error handling when the `name` field does not exist or does not have meaningful value. We surely do not want to remove anything than desired later.

pronan replied:
a little hard for me. wait

pronan replied:
`name` is a varible not a string literal, right? but I can't find where it is defined in `do_build`.

agentzh commented on bin/opm:
```diff
@@ -2195,6 +2198,37 @@ sub rebase_path ($$$) {
+            }
+        }
+        closedir $dh;
+    }else{
```
Ditto.

pronan replied:
done

agentzh commented on bin/opm:
```diff
@@ -2195,6 +2198,37 @@ sub rebase_path ($$$) {
+                }
+            }
+        }
+        closedir $dh;
```
Need error handling when `closedir` returns an error.

pronan replied:
done

agentzh commented on bin/opm:
```diff
@@ -2195,6 +2198,37 @@ sub rebase_path ($$$) {
+                }
+            }else{
+                if ($x =~ /^\Q$dist_name\E-[.\w]*\d[.\w]*\.tar\.gz$/){
+                    shell "rm -rf $x";
```
We should use the Perl builtin function `unlink` to remove a single file. `rm -rf` is overkill here.

pronan replied:
done

agentzh commented on bin/opm:
```diff
@@ -2195,6 +2198,37 @@ sub rebase_path ($$$) {
+        my $default_sec = $data->{default};
+        my $dist_name = $default_sec->{name};
+        opendir(my $dh, './') || die "Can't opendir";
+        my @dots = readdir($dh);
```
`dots` is a bad name. Maybe `entities` is better?

pronan replied:
done.

agentzh commented on bin/opm:
```diff
@@ -2195,6 +2198,37 @@ sub rebase_path ($$$) {
+        my $dist_name = $default_sec->{name};
+        opendir(my $dh, './') || die "Can't opendir";
+        my @dots = readdir($dh);
+        foreach my $x (@dots) {
```
`x` is a bad name. we need more descriptive names like `entity`, I think.

agentzh replied:
Oh, BTW, I think you can just use `for` here instead of `foreach`.

pronan replied:
done

agentzh commented on bin/opm:
```diff
@@ -2195,6 +2198,37 @@ sub rebase_path ($$$) {
+        }
+        closedir $dh;
+    }else{
+        err "invalid clean option: ".$_[0];
```
Again, we need spaces around the `.` operator. I think the error message `unrecognized argument for clean` is better. Also, we should give a list of recognized clean arguments here as well.

pronan replied:
done

agentzh commented:
Thanks for your work on this! Please address the issues in my latest round of review :) Thanks!

agentzh commented on bin/opm:
```diff
@@ -2195,6 +2198,37 @@ sub rebase_path ($$$) {
+        foreach my $x (@dots) {
+            if(-d $x){
+                if ($x =~ /^\Q$dist_name\E-[.\w]*\d[.\w]*$/){
+                    shell "rm -rf $x";
```
Oh, forgot to mention that, we also need to test if there is an `dist.ini` file in this directory before removing it recursively. `rm -rf` is scary and we can never be too cautious here ;)

pronan replied:
Do you mean if there is a `dist.ini` file in it, it should not be removed even if it matches the regex?

agentzh commented:
And...we also need to update the usage text in the program as well as the documentation for this new feature BTW :)

agentzh commented:
@pronan `$default_sec->{name}` is reading the data from the `dist.ini` file so it could be empty or bad. Just look at other similar places in the `opm` source file for examples of checking validity of this value.

pronan commented:
@agentzh Hello, Spring big brother, all done and test passed.

agentzh commented on bin/opm:
```diff
@@ -2195,6 +2198,57 @@ sub rebase_path ($$$) {
+            }
+        }
+        closedir $dh
+            or err "failed to close directory $dh!\n";
```
We usually do not use `!` at the end of our error string messages. Instead, you should use the special variable `$!` here to output the system error string for the failed `closedir` operation.

pronan replied:
done

agentzh commented on bin/opm:
```diff
@@ -2195,6 +2198,57 @@ sub rebase_path ($$$) {
+    my @valid_options = ('dist');
+    
+    if (@_ == 0) {
+        err "no clean option specified.\n";
```
I think "dist" is a clean *argument* instead of an option. In the CLI terminology, options are usually something like `--cwd` and `--help`.

pronan replied:
done

agentzh commented on bin/opm:
```diff
@@ -2195,6 +2198,57 @@ sub rebase_path ($$$) {
+        {
+            err "$dist_file: bad dist name: $dist_name\n";
+        }
+        opendir(my $dh, './') 
```
Need a blank line after the `if` statement block.

agentzh replied:
`./` looks strange to me. You can just use `.` here.

pronan replied:
done

pronan replied:
done

agentzh commented on bin/opm:
```diff
@@ -2195,6 +2198,57 @@ sub rebase_path ($$$) {
+            err "$dist_file: bad dist name: $dist_name\n";
+        }
+        opendir(my $dh, './') 
+            or err "failed to open directory './'!\n";
```
We need to use `: $!` instead of `!` here. See other similar places. Be careful ;)

pronan replied:
done

agentzh commented on bin/opm:
```diff
@@ -2195,6 +2198,57 @@ sub rebase_path ($$$) {
+                
+                if ($entity =~ /^\Q$dist_name\E-[.\w]*\d[.\w]*$/) {
+                
+                    if (-e "$entity/dist.ini") {
```
I think the `-f` test is stronger than `-e`. Let's use `-f` instead here.

pronan replied:
done

pronan replied:
done

agentzh commented on bin/opm:
```diff
@@ -2195,6 +2198,57 @@ sub rebase_path ($$$) {
+                
+                    if (-e "$entity/dist.ini") {
+                        shell "rm -rf $entity";
+                        print "delete folder: $entity\n";
```
"delete" => "removed", 
"folder" => "directory",
":" => ""

pronan replied:
done

agentzh commented on bin/opm:
```diff
@@ -2195,6 +2198,57 @@ sub rebase_path ($$$) {
+                
+                if ($entity =~ /^\Q$dist_name\E-[.\w]*\d[.\w]*\.tar\.gz$/) {
+                    unlink $entity;
+                    print "delete file: $entity\n";
```
"delete" => "removed"
": " => ""

pronan replied:
done

agentzh commented on bin/opm:
```diff
@@ -2195,6 +2198,57 @@ sub rebase_path ($$$) {
+        }
+
+        if (length $dist_name < 3
+            || $dist_name =~ /[^-\w]|^(?:nginx|luajit|resty|openresty|opm|restydoc.*|ngx_.*|.*-nginx-module)$/i)
```
Because this check is also used elsewhere. Maybe we should introduce a small Perl sub to avoid code duplication here? Like `sub is_valid_dist_name`?

pronan replied:
done

agentzh commented on bin/opm:
```diff
@@ -2195,6 +2198,57 @@ sub rebase_path ($$$) {
 }
 
+sub do_clean {
+    my @valid_options = ('dist');
```
We won't have many different kinds of arguments, especially when we only have one right now. No need to introduce a Perl array here. It is just referenced once and simply adds line noises IMHO :)

pronan replied:
done

agentzh commented on bin/opm:
```diff
@@ -2195,6 +2198,57 @@ sub rebase_path ($$$) {
+        closedir $dh
+            or err "failed to close directory $dh!\n";
+    } else {
+        err "unrecognized argument for clean: $_[0]. recognized clean arguments are: " . join(", ", @valid_options);
```
This line exceeds 80 columns a lot, I'm afraid ;)

pronan replied:
done

pronan commented:
Hello, chun ge. All done but I'm not sure how to merge my two PR so I did nothing for that.

agentzh commented:
@pronan You can just `git cherry-pick` the commit(s) in your branch for PR #26 over to your branch for this PR.

pronan commented:
@agentzh hello, Should I do this in local computer or just on github webpage? I clone my fork `pronan/opm` to my comupter, but I found it is same with yours. I do change two files. I don't know why.

agentzh commented:
@pronan You do `git cherry-pick` in your local branches and when your local branch is ready, you just `git push`. Unlike SVN, one seldom or never does such operations directly on the remote git branches.

agentzh commented:
@pronan Do not use GitHub's web UI to make changes. That tends to suck a lot.

pronan commented:
@agentzh Saddly, I was using github web UI to make changes.It sucks. Now I'm working on local branch.

agentzh commented:
cool.

agentzh commented:
This patch has been updated and merged with #26 to master.
